### PR TITLE
platform: add wirless powershare

### PR DESCRIPTION
**Pull request by Linux4**
Seems to work fine, please test before merging

### CODE REVIEW

**Comment by cryptomilk** [comment]
Can you explain what this is and how it works?

**Comment by Linux4** [comment]
See: https://www.samsung.com/global/galaxy/what-is/wireless-powershare/ This adds Support for wireless powershare on lineageos (like on ivanmeler's LOS). By writing 0/1 to /sys/class/power_supply/battery/wc_tx_en wireless powershare is disabled/enabled. It also adds a custom QS Tile to toggle wireless power share.

**Comment by cryptomilk** [comment]
Where do you have the implementation from?

File: config/init/init.exynos9820.rc [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import /vendor/etc/init/hw/init.exynos9820.root.rc
+import /vendor/etc/init/init.exynos9820.powershare.rc
```
Comment (cryptomilk) [inline config/init/init.exynos9820.rc]: Please put the init also in `/vendor/etc/init/hw/`

**Comment by Linux4** [comment]
I used strace to see what ivanmelers proprietary /system/bin/powersharedaemon does ... And saw it writing to /sys/class/power_supply/battery/wc_tx_en

**Comment by cryptomilk** [comment]
I mean the quick setting tile code :-)

**Comment by Linux4** [comment]
Looked up how to make custom quick settings tile (e.g [here](https://developer.android.com/reference/android/service/quicksettings/TileService)) (I have some experience developing Android Apps) but additionally looked into the decompiled apk found on ivanmeler's LOS.

**Comment by cryptomilk** [comment]
I dunno when I can test it. However the priority is to find the performance bug.

**Comment by derfelot** [comment]
pretty cool feature. not sure i have a device to test this though.

**Comment by cryptomilk** [comment]
I think this needs to be extened. For example it should be greyed out/not selectable if the battery saver is turned on.

See:
https://review.lineageos.org/c/LineageOS/android_frameworks_base/+/271613

Could you extend it please?

**Comment by cryptomilk** [comment]
According to the linked video it needs to be above 30% battery. So I guess we need to check the battery level before turning on.

**Review by cryptomilk — CHANGES_REQUESTED** [review]
Please add some saftey guards :-)

**Comment by Linux4** [comment]
> I think this needs to be extened. For example it should be greyed out/not selectable if the battery saver is turned on.
> 
> See:
> https://review.lineageos.org/c/LineageOS/android_frameworks_base/+/271613
> 
> Could you extend it please?

That's not as easy as it would be when the QS tile would be integrated into frameworks/base, will take some time to find out how to implement it without framework modifications.

**Comment by cryptomilk** [comment]
There is `/sys/class/power_supply/battery/wc_tx_stop_capacity` which tells you at how many % it should stop charging I guess.

Looks like you can write to it too. So this is the value when it should stop charging the other device.

**Comment by cryptomilk** [comment]
Then there is

```
WIRELESS_POWER_SHARING_SYSFS_EXTERNEL_EVENT_PATH = "/sys/class/power_supply/battery/ext_event";
```

 which takes an integer, but I'm not sure what this triggers.

**Comment by Linux4** [comment]
Now it should be disabled when you enable powersave, and when battery < 30%, but it seems to have some problems with SELinux as it can't read/set system props with selinux enforcing, can you help?

**Comment by Linux4** [comment]
wireless powershare is now working with selinux but the app still can't set the sys.powershare prop, set_prop(platform_app, vendor_powershare_prop) produces a neverallow error. Please also tell me when I did something wrong with the selinux rules.

**Comment by Linux4** [comment]
This needs to be implemented in LineageOS upstream to get it working with selinux enforcing.

**Pull request closed without merging**